requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,97 @@
+
+The scheduled changes can be stored on the flag record in the `scheduled_changes` field with the following structure:
+
+```
+[
+    {
+        "scheduled_at": "2023-11-22T12:30:00Z",
+        "type": "is_active",
+        "value": "false"
+    },
+    {
+        "scheduled_at": "2023-12-01T15:00:00Z",
+        "type": “filter”,
+        "value": {
+            "variant": null,
+            "properties": [{
+                "key": "$geoip_country_name",
+                "type": "person",
+                "value": ["Australia"],
+                "operator": "exact"
+            }],
+            "rollout_percentage": 20
+        }
+    }
+]
+```
+
+At the time specified, the Celery task fetches the flag, and iterates over the `scheduled_changes` items. For any item in this array, the `scheduled_at` field is evaluated, and if it is in the past, the task populates the item on the flag in the following way:
+
+-   When the type `is_active` the task sets the value of the `is_active` field.
+-   When the type is `filter`, the given condition is populated in the `filters` field of the flag.
+
+After populating the change on the feature flag, the change item is removed from the `scheduled_changes` array.
+
+To summarize, the end-to-end flow can go as follows:
+
+1. The user specifies the changes in the UI and saves the feature flag.
+2. The endpoint saves the record and stores a scheduled task in the Celery queue.
+3. Celery executes the scheduled task at the future date. It populates the required fields and reschedules the next task in the future.
+
+## UI
+
+We can add the functionality under the "Schedule" tab within the feature flag view. Users can submit scheduled changes via a form and view currently scheduled changes.
+
+<img width="981" alt="image" src="https://github.com/PostHog/posthog/assets/22996112/ff9f1c4c-c2f6-4dde-a420-bc282c5adcc8">
+
+## Generalizability
+
+The initial implementation idea assumes we schedule the task within the `POST /feature_flags` endpoint. However, we might want to abstract the scheduling functionality to accommodate other models such as `Survey` or `Experiment.` This would involve creating a scheduling API.
+
+```
+POST /projects/schedule
+{
+    type: 'feature_flag',
+    scheduled_changes: [ ... ]
+    scheduled_at: <timestamp>
+}
+```
+
+The API would populate the scheduled changes on the record and store a task in the Celery queue.
+
+I considered the option of only storing the scheduled changes in the task itself. This is, however, problematic since we need the list of changes available on the flag for the table view in the UI. That means this API would still need to both store the task and update the record with the list of scheduled changes.
+
+## Possible failure modes
+
+-   When submitting the scheduled changes to perform in the future, we need to store both the list of changes on the flag and the Celery task. The order matters here - we want to store the _task_ first. This is because:
```

Comment:
Having two sources of truth is indeed problematic, and I don't think we should go down this route.

We should use the scheduling only as a trigger, i.e. it's time to do something, and the 'something' has its own logic for figuring out what to do. More concretely this can look something like:

1. We have a table of all scheduled changes for a model, with the scheduled_at time, and a payload
2. We can use this table to show scheduled changes for a given model-item
3. Celery triggers a job every minute that checks for any model with scheduled_at <= now & if found, runs it & deletes / marks completed.
4. We use the db transaction guarantees to make sure it's executed by only one worker at a time.

I think it's worth generalising _right now_ rather than building something on the flag model itself.

The part I don't 100% like about this is we're building our own message queue, and all the issues that come with it. It's _fine if that's what works best_, but want us to consciously decide so.